sys-fs/e2fsprogs/files/fsck_ext2fs.8
.TH fsck_ext2fs 8 2006-07-02 "Matthias Andree" "FreeBSD Ports"
.\"
.\" fsck_ext2fs.8 - manual page for fsck_ext2fs wrapper
.\"
.\" (C) Copyright 2006 by Matthias Andree <matthias.andree@gmx.de>
.\"
.\" License: This file may be redistributed in accordance with the terms
.\" of the GNU General Public License v2.
.\"
.\" Upstream $Id: fsck_ext2fs.8,v 1.3 2006/07/02 11:45:21 emma Exp $
.\" $FreeBSD: ports/sysutils/e2fsprogs/files/fsck_ext2fs.8,v 1.1 2006/07/04 15:47:51 leeym Exp $
.\"
.SH NAME
.B fsck_ext2fs
\- compatibility wrapper for e2fsck
.SH SYNOPSIS
.P
.B fsck_ext2fs
[\fB\-Fpfnyv\fR] [\fB\-b\fR \fIblock\fR]
.SH DESCRIPTION
.P
\fBfsck_ext2fs\fR maps the traditional FreeBSD \fBfsck_ffs\fR options to
options with the same functionality for \fBe2fsck,\fR runs \fBe2fsck\fR
and then maps its exit status to values that FreeBSD understands.
\fBe2fsck\fR is a utility to check and repair ext2 and ext3 file
systems.

.SH OPTIONS
.IP \fB\-F\fR
(check foreground mode required) Immediately exits with status 1 to tell
\fBfsck\fR that ext2fs cannot be checked in the background. \fBfsck\fR
usually runs \fBfsck_*\fR programs twice, first with \fB\-F\fR to find
out if they can do background checking, then either immediately without
\fB\-F\fR for foreground checking or deferred in the background with
\fB\-B\fR.
.IP \fB\-p\fR
(preen mode) This option suppresses adding the \fB\-f\fR option (unless
\fB\-f\fR is also given) and adds the \fB\-p\fR option to the
\fBe2fsck\fR command line. This causes \fBe2fsck\fR to automatically fix
any filesystem problems that can safely be fixed without operator
intervention. Without this option given, \fBe2fsck\fR will be run with
the \fB\-f\fR option to force a check, since interactive scan and repair
mode is the default on FreeBSD, but not on Linux where \fBe2fsck\fR
comes from.
.IP \fB\-f\fR
(force check) This option forces the check of a clean file system while
preening and is passed to \fBe2fsck\fR verbatim.
.IP \fB\-n\fR
("no" mode) This option causes the file system to be opened in read-only
mode and assume "no" as answer to all questions. This is the only way to
safely run \fBfsck\fR on a mounted ext2 or ext3 file system. This option
is passed to \fBe2fsck\fR verbatim.
.IP \fB\-y\fR
("yes" mode) This option is passed verbatim to \fBe2fsck\fR and causes
it to assume "yes" as answer to all questions. This allows the
non-interactive use of e2fsck but is rather aggressive. Use with care.
.IP \fB\-v\fR
(verbose output) This option is passed verbatim to \fBe2fsck\fR and
causes it to verbosely report its progress.
.IP "\fB\-b\fR \fIblock\fR"
(use alternate super block) This option is passed verbatim to
\fBe2fsck\fR and selects an alternate super block, for use when the
primary super block has been damaged. Please see the \fBe2fsck\fR(8)
manual page for details.

.SH EXIT STATUS
If errors remain after \fBe2fsck\fR, an invalid option or too many
options have been specified, \fBe2fsck\fR was killed with a signal or
the \fIfork\fB system call failed, \fBfsck_ext2fs\fR exits with status
EXIT_FAILURE (usually 1). If \fBe2fsck\fR cannot be started, exits with
status 127. If the file system is clean after \fBe2fsck\fR operation,
exits with status EXIT_SUCCESS (0).

.SH NOTES
.P
This utility is merely meant as an adaptor so that \fBe2fsck\fR can be
run during the boot process, it does not support all options that
\fBe2fsck\fR offers. If you need one of its advanced options, please run
\fBe2fsck\fR directly.

.SH FILES
.TP
.I /sbin/e2fsck
is the location of the \fBe2fsck\fR program to run.

.SH AUTHOR
.P
Matthias Andree <matthias.andree@gmx.de> wrote the program and this
manual page.
.SH CONFORMING TO
The FreeBSD 6.1 command line interface for \fBfsck_ufs\fR(8).
.SH SEE ALSO
.BR fsck (8),
.BR e2fsck (8)
and
.BR fsck_ufs (8).


eclass/ChangeLog
# ChangeLog for eclass directory
# Copyright 1999-2015 Gentoo Foundation; Distributed under the GPL v2
# $Header: /var/cvsroot/gentoo-x86/eclass/ChangeLog,v 1.1750 2015/08/08 09:32:50 mgorny Exp $

  08 Aug 2015; Michał Górny <mgorny@gentoo.org> git-r3.eclass:
  Add some boldness to output. Update/fix pkg_needrebuild() for
  smart-live-rebuild.

  07 Aug 2015; Davide Pesavento <pesa@gentoo.org> qt5-build.eclass:
  Fix bugs #549140 and #552942.

  07 Aug 2015; Davide Pesavento <pesa@gentoo.org> qt5-build.eclass:
  Allow passing arguments to qmake from ebuilds via the myqmakeargs array.

  06 Aug 2015; Sergei Trofimovich <slyfox@gentoo.org> ghc-package.eclass,
  haskell-cabal.eclass:
  Add new helpers: 'ghc-pm-version' to get ghc version as seen by package
  manager and 'ghc-is-dynamic' to workaround ghc-api bug
  https://ghc.haskell.org/trac/ghc/ticket/10301 in ebuild.

  05 Aug 2015; William Hubbs <williamh@gentoo.org> golang-vcs-snapshot.eclass:
  add || die and fix indentation

  05 Aug 2015; William Hubbs <williamh@gentoo.org> +golang-vcs-snapshot.eclass:
  Add eclass for vcs snapshots of software written in Go.

  03 Aug 2015; Michael Sterrett <mr_bones_@gentoo.org> python-utils-r1.eclass:
  Add quotes to support reading from files with spaces in the filename.

  03 Aug 2015; Michael Palimaka <kensington@gentoo.org> kde5.eclass:
  Do not compress handbooks.

  02 Aug 2015; James Le Cuirot <chewi@gentoo.org> java-utils-2.eclass:
  Allow java-pkg_get-javac to be called alone. Fixes bug #172594. Thanks to
  sping for the patch. Also simplify die logic around java-pkg_javac-args. die
  handling has improved since this was written.

  31 Jul 2015; Patrice Clement <monsieurp@gentoo.org> java-utils-2.eclass:
  Document JAVA_RM_FILES variable. Also, fix documentation and correct typos in
  java-pkg_rm_files.

  30 Jul 2015; James Le Cuirot <chewi@gentoo.org> java-utils-2.eclass:
  Fix JAVA_RM_FILES by simplifying java-utils-2_src_prepare on the basis that
  we won't be committing EAPI 0|1 ebuilds any more.

  30 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-multilib.eclass:
  Revert bad mycmakeargs changes and introduce 3 eclass variables to have extra
  defines in the ebuild

  30 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-multilib.eclass:
  Reset variables between ABIs

  30 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-multilib.eclass:
  Add blocker below virtual/mysql-5.6-r4 for new split ebuilds

  29 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-multilib.eclass:
  Fix pkg_config function error with non-existant USE flags

  28 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-multilib.eclass:
  Fix configuring non-native abi which leads to build failure wrt bug 556162

  28 Jul 2015; Brian Evans <grknight@gentoo.org> mysql-cmake.eclass,
  mysql-multilib.eclass:
  Add support for the split client/server options to the mysql eclasses

  27 Jul 2015; William Hubbs <williamh@gentoo.org> +golang-base.eclass,
  golang-build.eclass, golang-vcs.eclass:
  Add golang-base.eclass for the basic golang functions and set up the
  other go eclasses to use it.


  27 Jul 2015; Michał Górny <mgorny@gentoo.org> python-utils-r1.eclass:
  python_wrapper_setup(): make banned helpers exit with 127.

  27 Jul 2015; Michał Górny <mgorny@gentoo.org> python-utils-r1.eclass:
  Ban calling pythonN and pythonN-config when the other version of Python is
  selected (i.e. ban python2 when python3 is used).

  27 Jul 2015; Michał Górny <mgorny@gentoo.org> python-utils-r1.eclass:
  python_wrapper_setup(): wrap pythonN-config as well as suggested by PEP and
  required by some applications, bug #555594.

  27 Jul 2015; Michał Górny <mgorny@gentoo.org> python-utils-r1.eclass:
  python_wrapper_setup(): replace symlinks with shell wrappers to avoid
  triggering Python 3.4+ magical prefix support.

  27 Jul 2015; Mike Gilbert <floppym@gentoo.org> python-r1.eclass:
  Drop the USE_PYTHON warning.

  25 Jul 2015; Michał Górny <mgorny@gentoo.org> python-utils-r1.eclass:
  Add missing ||die to "rm -f" calls, i.e. in case we do not have permission to
  remove the files.

  23 Jul 2015; William Hubbs <williamh@gentoo.org> golang-build.eclass:
  Add functions to retrieve Go library paths and install Go packages.

  22 Jul 2015; Michael Palimaka <kensington@gentoo.org> kde4-meta.eclass:
  Fix oldpim unpack by Andreas Sturmlechner <andreas.sturmlechner@gmail.com>
  wrt bug #555566.

  22 Jul 2015; Patrice Clement <monsieurp@gentoo.org> java-utils-2.eclass:
  Introduce java-pkg_rm_files as a helper function along with JAVA_RM_FILES
  array to readily get rid of useless files.

  20 Jul 2015; Sergei Trofimovich <slyfox@gentoo.org> haskell-cabal.eclass:
  Workaround upstream cabal tests hangup bug #537500 by Michael Orlitzky; use
  ghc's haddock for doc generation.

  19 Jul 2015; Mike Pagano <mpagano@gentoo.org> kernel-2.eclass:
  Change kernel upgrade http link and remove reference to deblob in elog
  message. See bug #553484

  19 Jul 2015; Johannes Huber <johu@gentoo.org> kde4-base.eclass:
  Cleanup SRC_URIs.

  18 Jul 2015; Mike Gilbert <floppym@gentoo.org> distutils-r1.eclass:
  Add entire python directory to SANDBOX_PREDICT, bug 554252.

  17 Jul 2015; Ian Stakenvicius (_AxS_) <axs@gentoo.org>
  mozconfig-v6.38.eclass, mozconfig-v6.39.eclass:
  Updated gstreamer dependencies to current stable and dropped gst-plugins-mad
  since gst-plugins-libav now officially provides mp3 decoding

  17 Jul 2015; Brian Evans <grknight@gentoo.org> depend.php.eclass:
  Remove deprecated functions from depend.php.eclass as announced 30 days ago

  17 Jul 2015; Brian Evans <grknight@gentoo.org> -php-common-r1.eclass,
  -php-ext-base-r1.eclass, -php-ezc.eclass:
  Drop old, unused eclasses wrt bug 551910

  17 Jul 2015; Ian Stakenvicius (_AxS_) <axs@gentoo.org>
  mozconfig-v6.38.eclass, mozconfig-v6.39.eclass:
  Added gst-plugins-mad:1.0 dependency for USE=gstreamer to ensure mp3
  streaming support


sys-fs/e2fsprogs/files/fsck_ext2fs.c
/*
 * fsck_ext2fs - wrapper for e2fsck on FreeBSD
 * Copyright (C) 2004,2006 Matthias Andree <matthias.andree@gmx.de>
 * redistributable in accordance with the
 * GNU General Public License v2
 *
 * $FreeBSD: ports/sysutils/e2fsprogs/files/fsck_ext2fs.c,v 1.5 2006/07/04 15:47:51 leeym Exp $
 *
 * Upstream: $Id: fsck_ext2fs.c,v 1.6 2006/07/02 11:37:49 emma Exp $
 *
 * format: gindent -kr
 */

#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <errno.h>

__attribute__ ((noreturn))
static int die(const char *tag)
{
	perror(tag);
	exit(EXIT_FAILURE);
}

int main(int argc, char **argv)
{
	int ch, i = 1, force = 0, status, verbose = 0, t;
	long block = 0;
	enum { normal, preen, yes, no } mode = normal;
	char *cmd[256];
	pid_t pid;

	cmd[0] = "/sbin/e2fsck";
	while ((ch = getopt(argc, argv, "BFpfnyb:v")) != -1) {
		switch (ch) {
		case 'p':
			mode = preen;
			break;
		case 'f':
			force = 1;
			break;
		case 'n':
			mode = no;
			break;
		case 'y':
			mode = yes;
			break;
		case 'b':
			block = atol(optarg);
			break;
		case 'v':
			verbose++;
			break;
		case 'F':
			/* e2fsck does not support background checking,
			 * hence exit with nonzero status to force
			 * the foreground check. */
			exit(1);
		case 'B':
		default:
			fprintf(stderr, "%s: unknown option -%c\n",
				argv[0], optopt);
			exit(EXIT_FAILURE);
		}
	}

	if (force)
		cmd[i++] = "-f";

	switch (mode) {
	case normal:
		/* FreeBSD needs -f to force a check only in context
		 * with -p -- so map normal to force to match
		 * expectations */
		if (!force)
		    cmd[i++] = "-f";
		break;
	case yes:
		cmd[i++] = "-y";
		break;
	case no:
		cmd[i++] = "-n";
		break;
	case preen:
		cmd[i++] = "-p";
		break;
	}

	if (block) {
		static char b[30];

		sprintf(b, "-b %ld", block);
		cmd[i++] = b;
	}

	/* silently limit verbose to 15 so we don't overflow the cmd array */
	if (verbose > 15)
	    verbose = 15;

	for (t = verbose; t > 1; t--)
	    cmd[i++] = "-v";

	while (optind < argc) {
		cmd[i++] = argv[optind++];
		/* sanity check so we don't overflow the cmd buffer */
		if (i+1 == sizeof(cmd)/sizeof(cmd[0])) {
		    errno = E2BIG;
		    die(argv[0]);
		}
	}

	cmd[i++] = 0;

	if (verbose) {
		for (i=0; cmd[i]; i++)
			fputs(cmd[i], stderr),
			fputc(' ', stderr);
		fputc('\n', stderr);
	}

	pid = fork();
	switch (pid) {
	case -1:
		/* error */
		die("fork");
		break;
	case 0:
		/* child */
		(void) execv(cmd[0], cmd);
		perror("execve");
		_exit(127);
	default:
		/* parent */
		if (pid != waitpid(pid, &status, 0))
			die("waitpid");
		if (WIFSIGNALED(status)
		    || (WIFEXITED(status) && WEXITSTATUS(status) >= 4))
			exit(EXIT_FAILURE);
	}
	exit(EXIT_SUCCESS);
}
